### DISCUSSION TITLE
Deepseek V4.1 Flash on H100

### DISCUSSION THREAD

**Discussion by mtc-20** [sgl-project/sglang#39791] [General]
Has anyone been able to run the Deepseek V4.1 Flash on H100s? 
I tried using  the configs from the cookbook (I used the H200 template, then switched the backend to marlin to see if there was a difference), but the throughput (for single requests) was in the 40-50 tokens/s range, whereas the Deepseek V4 Flash 0731 on the same GPUs nets me 150+ tokens/sec. For larger prompts or even 2 parallel requests, the server crashes due CUDA OOM.

I tried deploying via vllm, and there for the same set of prompts, the throughput is in the 120+ range.

I'm curious if anyone else has been able to deploy via sglang on H100s, and what throughput you've been getting?

I would really like to have this deployed and running on H100s

**Comment by dcongthinh0-ship-it** [comment]
我的4卡h100就50多out throughput，

**Comment by KushagraKanaujia** [comment]
@mtc-20 The cookbook has no V4.1 Flash cell. Its H200 Flash recipes run --tp 4, while the H100 cells use --tp 8 with --moe-runner-backend marlin, so too low a TP on H100s could explain the OOM.

Comparing engine configs honestly is the problem I built Throttle for: it won't call a winner until repeat runs have measured noise.

```sh
throttle check --url http://localhost:30000 --model <model> --gpu-hourly-rate <rate> --concurrency 1 --label sglang-h100
```

https://github.com/KushagraKanaujia/throttle

I'll set it up with you on your box, free.

**Comment by flymysql** [comment]
## The short answer: H100 is not a supported target for V4.1 Flash, and the cookbook says why

Two things in this thread are worth correcting before you spend more GPU hours, plus the memory math that explains your OOM.

### 1. V4.1 Flash is not in any SGLang release yet

From the model's own cookbook page (`docs/cookbook/autoregressive/DeepSeek/DeepSeek-V4_1.mdx`):

> *"DeepSeek-V4.1 **Flash support has not shipped in an SGLang release yet**, so a stock `pip install sglang` cannot serve it. Use the preview Docker image below."*
> *"every NVIDIA cell on this page uses the preview build `lmsysorg/sglang:dev-dsv41` (H200 / B200 / B300 / GB300)"*

So if you're on `lmsysorg/sglang:latest` or a pip install, you are not running a build that is validated for this model at all. That alone can account for a throughput number that doesn't match vLLM's.

### 2. H100 is not in the supported hardware list for this model

The recipe definition is explicit:

```js
// docs/src/snippets/configs/deepseek-ai/deepseek-v4_1.jsx
supportedHardware: ["h200", "b200", "b300", "gb300", "mi350x"],
```

**`h100` is absent.** Compare with the previous generation, where H100 cells do exist and are marked verified:

```js
// deepseek-v4.jsx — V4 (not V4.1), H100 Flash
{ match: { hw: "h100", variant: "flash", quant: "fp4", strategy: "low-latency", nodes: "single" },
  verified: true,
  flags: [ "--model-path {{MODEL_NAME}}", "--tp 8",
           "--moe-runner-backend marlin",
           "--speculative-algorithm EAGLE", ... ] }
```

That difference is exactly why your V4-Flash-0731 baseline hits 150+ t/s while V4.1 Flash doesn't: **the older model has a validated H100 recipe; V4.1 does not.**

### 3. The H200 cell is the tightest one in the whole cookbook — and it's TP8+EP8

This is the part that explains your OOM, and where the other reply has it backwards. The comment above the H200 cell reads:

> *"H200: 8x H200, **TP8 + EP8**. No MXFP8 dense path on Hopper. **The tightest board here at 140 GiB**, so both cells also need the memory fraction pulled back; the cap alone still leaves the graphs short."*

and the flags are:

```
--tp 8 --ep-size 8 --mem-fraction-static 0.8
--attention-backend dsv4
--moe-runner-backend flashinfer_mxfp4
--cuda-graph-max-bs-decode 64
```

So for V4.1 the H200 recipe is **TP8**, not TP4 — the reply's "the H200 Flash recipes run `--tp 4`" is reading the *V4* table (which does say `H200 (TP=4)`) and applying it to V4.1. Being on TP8 on the *tightest supported board* is the relevant fact: **H100-80GB is strictly worse than the tightest board SGLang supports for this model.**

### 4. The memory math behind your OOM

The checkpoint is **475 GiB** (48 safetensors, fp8 dense weights + fp4 routed experts):

```
deepseek-ai/DeepSeek-V4.1-Flash   total safetensors = 510,296,708,312 bytes = 475.25 GiB
```

On 8×H100-80GB you have ~640 GB of HBM total. Weights alone are ~510 GB, which is **~80% of the node** before a single KV page, activation, CUDA-graph pool, or the DSpark draft head. That's why:
- single requests crawl (you're swapping/contending, not compute-bound), and
- *"larger prompts or even 2 parallel requests"* OOM — there is essentially no KV headroom left.

On 8×H200 (141 GiB = 1128 GB) the same 510 GB is ~45%, which is why the H200 cells work with `--mem-fraction-static 0.8`. This is a capacity problem, not a tuning problem: no flag combination fits a 475 GiB checkpoint comfortably into 640 GB.

### 5. There's also a real Hopper path penalty, independent of memory

Even when it runs, Hopper takes a degraded sparse-attention path:

```python
# python/sglang/srt/layers/attention/deepseek_v4_backend.py
@functools.lru_cache(maxsize=None)
def _is_sm100_or_newer() -> bool:
    # DeepGEMM's fp8_fp4 mqa-logits kernels need SM100+; Hopper takes the torch indexer.
    return torch.cuda.get_device_capability()[0] >= 10
```

V4.1 Flash is a **sparse-attention** MoE (each retrieving layer picks a top-512 candidate set from a compressed-latent store), so the indexer is squarely on the critical path. On SM90 that flag is `False`, and the fused DeepGEMM fp8_fp4 mqa-logits kernels / fused index store are all gated off in favour of a torch indexer. That's a plausible chunk of your 40–50 vs 120+ gap, separate from the OOM.

### 6. On "switching the backend to marlin"

Be careful — that's the V4 (previous-gen) H100 recipe, not a V4.1 knob. V4.1's own recipes specify `--attention-backend dsv4` + `--moe-runner-backend flashinfer_mxfp4` and the cookbook says:

> *"`--attention-backend`, `--moe-runner-backend` and `--fp8-gemm-backend` are selected automatically from the model, hardware, forward mode and shape. ... **Confirm them in the startup log rather than passing them.**"*

So the first useful datapoint from your run is the startup log's resolved triple. If it selected `dsv4` and then degraded internally on SM90, that's visible there.

### What I'd actually do

1. **Confirm the image.** You need `lmsysorg/sglang:dev-dsv41`; anything else is unvalidated for V4.1.
2. **Accept that H100 is out of scope for now** for the full model, and either use ≥H200-class hardware or keep running **V4 Flash 0731 on H100s**, which *does* have a verified H100 recipe (`--tp 8 --moe-runner-backend marlin`, `verified: true`). Your 150+ t/s baseline is the right thing to be comparing against, and it's a supported configuration.
3. If you want to push on H100 anyway, the honest experiments are: TP16 across 2 nodes (the V4 Pro H100 cells use `--tp 16` + `SGLANG_SHARED_EXPERT_TP1=1`), and `SGLANG_ENABLE_DSV41_ENGRAM_HOST_TABLE=1`, which the cookbook notes **frees HBM by moving the two fp8 Engram tables to host and removes the per-Engram-layer all-reduce** — that's the one documented lever that buys you back device memory without changing output. It costs host RAM, a longer load, and needs huge-page backing.
4. Post the startup-log triple plus your `max_total_num_tokens` and I can tell you whether you're purely capacity-bound or also losing to the indexer fallback.

One caveat on my own claims: everything above is from the current `main` cookbook and source, not from a run on H100s — I haven't executed this on your hardware. But the `supportedHardware` list, the TP8+EP8 H200 recipe, the 475 GiB checkpoint size, and the `_is_sm100_or_newer()` gate are all direct reads, and together they predict exactly the symptoms you're describing.

*Verified against `main` on 2026-09-25: `docs/cookbook/autoregressive/DeepSeek/DeepSeek-V4_1.mdx` (release status, hardware list, DSpark/Engram/SWA notes), `docs/src/snippets/configs/deepseek-ai/deepseek-v4_1.jsx` (`supportedHardware`, H200 TP8+EP8 cell), `docs/src/snippets/configs/deepseek-ai/deepseek-v4.jsx` (V4 H100 cells, `verified: true`), `python/sglang/srt/layers/attention/deepseek_v4_backend.py` (`_is_sm100_or_newer`), and the HF API for checkpoint size.*

**Reply by KushagraKanaujia** [reply]
Good catch, @flymysql, thanks. I read the V4 table (H200 TP=4) and applied it to V4.1; the V4.1 recipe is TP8+EP8 and H100 isn't a supported target. @mtc-20, go with flymysql's answer.
